staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -323,6 +327,10 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
 	}
 
 	if err := cmd.Run(); err != nil {
+		if a.installHint != "" {
+			return fmt.Errorf("exec: %v\n%s", err, a.installHint)
```

Comment:
This looks like it could produce a lot of log spam in the worst case scenario. Should we ratelimit the hint or only return the full hint on the first error?

Reply:
Mmm, fair. I am leaning towards ratelimiting the hint (exponential backoff?) because of this use case:
- client-go consumer gets into crash loop
- User sees that client-go consumer is in crash loop a non-trivial amount of time after the crash loop started
- User checks the logs the for client-go consumer
- User still wants to see the install hint, and doesn't want to have to scroll all the way to the first instance of the crash, which may be a long way up the log or in another file (in the case of log rotation)

1) Do you agree with the use case?
2) If so are you amenable to using an exponential backoff for printing the installHint? Something like print it the 1st time, print it the 2nd time, print it the 4th time, - 8th -, - 16th -, ..., - 128th -, - 256th -, - 256th -, - 256th -, ... Can you think of anywhere else in the codebase we do this?

Reply:
Suggest doing time interval based without backoff. Something simple like:

```golang
type Sometimes struct {
	Interval time.Duration

	mu    sync.Mutex
	last  time.Time
}

func (s *Sometimes) Do(f func()) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if !s.last.IsZero() && time.Since(s.last) >= s.Interval {
		f()
		s.last = time.Now()
	}
}
```

Reply:
Gave this a go in https://github.com/kubernetes/kubernetes/pull/91305/commits/dacfdc56d4caa8cdb20488609d4258a7f3576a60.

Reply:
Ah! Now (as of [0]) when I build `kubectl` and run `kubectl get pods` with an exec provider that uses an unknown binary, I don't see the `InstallHint` printed out anymore! It is just the short version[1] of the error. I gotta look into this (tomorrow) :(. Seems like `kubectl` is calling the exec provider multiple times?

```
0xc000888120.wrapCmdRunErrorLocked(exec: "does not exist": executable file not found in $PATH)
0xc000888120.wrapCmdRunErrorLocked(exec: "does not exist": executable file not found in $PATH)
0xc000888120.wrapCmdRunErrorLocked(exec: "does not exist": executable file not found in $PATH)
0xc000888120.wrapCmdRunErrorLocked(exec: "does not exist": executable file not found in $PATH)
0xc000888120.wrapCmdRunErrorLocked(exec: "does not exist": executable file not found in $PATH)
Unable to connect to the server: getting credentials: exec: executable does not exist not found
```

[0] e53aca0c6 (Improve exec binary unknown/failed error messages)
[1] https://github.com/ankeesler/kubernetes/commit/e53aca0c6f8c9a2c82abafd2e7baf6bd0b1c894e#diff-0861e1cd492e078057f9ec7524d9d6ffR415

Reply:
What if we had a window of time that the hint was printed in? I.e., the `InstallHint` will be printed every `1h` for `3s`. This seems to solve the log spew problem and the above problem with the `InstallHint` not showing up via `kubectl` (retries?).

This has the added benefit that there is a wider target (i.e., instances of the `InstallHint` in a log) for a user to find out what is going on w.r.t. their exec plugin.

Further typing out loud - this seems janky, because if `kubectl` hangs setting up a networking connection for too long, then it will defeat the purpose of the change. :(

Reply:
Could we have a threshold for the number of times that the verbose error message is printed? Say, the error message gets printed 10 times, and then it shuts up for the next hour, and then it prints 10 times again?

Again, seems super janky. I am wondering if there are other examples of this in the codebase...or any other codebase, for that matter.

Reply:
So to summarize potential paths forward, in order of least worst idea to most worst idea:
1) A better idea that someone else has :).
2) Have a threshold for number of times the verbose error gets printed in a window (code seems like it is more complicated than it is worth).
3) Log the verbose error with `klog.V(1)` (90% of `kubectl` users won't see this (assuming that 90% of `kubectl` users don't set their log level), still may result in log spew).

I'm leaning towards #2, since I can't think of anything else. :(

Reply:
Tried to do 2 here: https://github.com/kubernetes/kubernetes/pull/91305/commits/66876c589896a2557bb41b7f4f93e18fba222491 (Rate limit install hint printing by time and count).